floyd/cli/auth.py

```diff
@@ -9,22 +11,125 @@
+    def __init__(self, server_address, RequestHandlerClass, key_queue):
+        super(LoginServer, self).__init__(server_address, RequestHandlerClass)
+        self.key_queue = key_queue
+
+
+class LoginHttpRequestHandler(BaseHTTPRequestHandler):
+    def do_OPTIONS(self):
+        self.send_response(200, "ok")
+        self.send_header('Access-Control-Allow-Origin', '*')
+        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
+        self.end_headers()
+
+    def do_GET(self):
+        params = urlparse.parse_qs(urlparse.urlparse(self.path).query)
+        key = params.get('apikey')
+        if not key:
+            self.send_response(400)
+            return
+
+        self.server.key_queue.put(key[0])
+
+        self.send_response(200)
+        self.send_header('Access-Control-Allow-Origin', '*')
+        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
+        self.send_header('Content-type', 'text/html')
+        self.end_headers()
+        self.wfile.write("OK".encode('utf-8'))
+
+    def log_message(self, fmt, *args):
+        return
+
+
+@contextlib.contextmanager
+def get_free_port():
+    try:
+        s = socket.socket()
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        s.bind(('localhost', 0))
+        yield s.getsockname()[1]
+        s.close()
+    except socket.error:
+        yield None
+
+
+def wait_for_apikey():
+    key_queue = Queue()
+    with get_free_port() as port:
+        if not port:
+            return
+        server = LoginServer(('', port), LoginHttpRequestHandler, key_queue)
+
+    t = threading.Thread(
+        target=server.serve_forever)
+    t.daemon = True
+    t.start()
+
+    cli_host = 'http://127.0.0.1'
+    url = '%s/cli_login?callback=%s:%s' % (floyd.floyd_web_host, cli_host, port)
+    subprocess.check_output(
+        [sys.executable, '-m', 'webbrowser', url], stderr=subprocess.STDOUT)
+
+    floyd_logger.info('waiting for login from browser...')
+    while True:
+        try:
+            apikey = key_queue.get(timeout=500)
+            break
+        except Queue.Empty:
+            pass
+
+    server.shutdown()
+    return apikey
+
+
+def has_browser():
+    try:
+        webbrowser.get()
+        return True
+    except webbrowser.Error:
+        return False
+
 
 @click.command()
 @click.option('--token', is_flag=True, default=False, help='Just enter token')
-@click.option('--apikey', '-k', default=None, help='Api Key')
-@click.option('--username', '-u', default=None, help='FloydHub username')
-@click.option('--password', '-p', default=None, help='FloydHub password')
+@click.option('--apikey', '-k', help='Api Key')
+@click.option('--username', '-u', help='FloydHub username')
+@click.option('--password', '-p', help='FloydHub password')
 def login(token, apikey, username, password):
     """
     Login to FloydHub.
     """
+    if not apikey:
+        if has_browser():
+            apikey = wait_for_apikey()
```

Comment:
Is it possible to exit the `wait_for_apikey` call and continue this function? So that the user could login with token or username / pwd?